lwt/cohttp_lwt.ml

```diff
@@ -194,21 +210,32 @@ module Make_client
   let callv ?(ctx=default_ctx) uri reqs =
     lwt (conn, ic, oc) = Net.connect_uri ~ctx uri in
     (* Serialise the requests out to the wire *)
-    let _ = Lwt_stream.iter_s (fun (req,body) ->
-        Request.write (fun writer ->
-            Cohttp_lwt_body.write_body (Request.write_body writer) body) req oc)
-        reqs in
-    (* Read the responses. For each response, ensure that the previous response
-     * has consumed the body before continuing to the next response, since HTTP/1.1
-     * pipelining cannot be interleaved. *)
+    lwt meths = Lwt_stream.fold_s (fun (req,body) meths ->
+      Request.write (fun writer ->
+        Cohttp_lwt_body.write_body (Request.write_body writer) body
+      ) req oc >>= fun () ->
+      return ((Request.meth req)::meths)
```

Comment:
I chose the fold to make sure the requests were actually written. I wasn't clear on the meaning of `Lwt_stream.map_s` because it seems like it simply records the mapping but doesn't force its application. The `fold_s` looked like the functional equivalent of the original `iter_s`.

Maybe this is OK but then a request may not be written to the wire until its response is trying to be read which sort of defeats the purpose of pipelining.

Basically, it comes down to ensuring the correct pipelining behavior. If there is a way to eagerly consume requests and lazily consume responses, we should use that. Perhaps we need some kind of "pump" lwt thread between the request stream and response stream to ensure that requests are dispatched ASAP and response consumption does not "pull" requests out to the wire?